bibtasklets/bst_orcidsync.py

```diff
@@ -44,24 +45,74 @@
 from invenio.dbquery import run_sql
 from invenio.webuser import collect_user_info
 from invenio.bibtask import write_message
+from invenio.bibauthorid_webapi import get_hepnames
 
 CFG_REDIS_ORCID_SYNC_KEY = 'legacy_orcid_tokens'
 CFG_REDIS_ORCID_SYNC_KEY_COUNT = 'legacy_orcid_tokens_count'
 
 
 def orcid_data_to_sync():
-    """Returns an iterator of data to sync."""
+    """
+    Collect all available OAuth tokens for ORCID and their related data:
+    (orcid, token, email, name)
+    Orcid and token are mandatory data.
+    Email and name can be empty strings.
+    In order to collect the email and the name, the strategy is:
+    - if a uid is found, then try to get them from the user table.
+    - if the email or the name are empty, then try to get them from HepNames.
+
+    Note regarding account types.
+    There are 2 different types of account involved.
+    1. The Inspire user account that requires an email and name and assigns
+    a uid (in particular when users log in via arXiv). The BAI system personid
+    is separate and independent of the Inspire user accounts uid, but if a
+    connection can be made to a user account the uid is associated with the
+    personid in the aidPERSONIDDATA table.
+    2. HepNames which has personal data like name and email and is associated
+    with the BAI via MARC: 000996436 035__ $$9BAI$$aSimona.Murgia.1
+    There are some convenience methods to get info from HepNames:
+    from invenio.bibauthorid_webapi import get_hepnames
+    """
     rows = run_sql("SELECT personid, data, token FROM aidPERSONIDDATA NATURAL JOIN aidTOKEN WHERE tag='extid:ORCID';")
     for personid, orcid, token in rows:
+        orcid = orcid.decode('utf8')
+        token = token.decode('utf8')
+        email = u''
+        name = u''
+
         uid = run_sql("SELECT data FROM aidPERSONIDDATA WHERE personid=%s AND tag='uid'", (personid, ))
+
+        # If uid, then collect name and email from the user table.
         if uid:
             uid = int(uid[0][0])
             user_info = collect_user_info(uid)
+            email = _validate_email(user_info['email'].decode('utf8'))
             if user_info.get('external_firstname') and user_info.get('external_familyname'):
                 name = u'%s %s' % (user_info['external_firstname'].decode('utf8'), user_info['external_familyname'].decode('utf8'))
             else:
                 name = user_info['nickname'].decode('utf8')
-            yield orcid.decode('utf8'), token.decode('utf8'), user_info['email'].decode('utf8'), name
+
+        # If name or email are empty, then try to collect them from HepNames.
+        if not email or not name:
+            hepnames = get_hepnames(personid)
+
+        if hepnames and not email:
```

Comment:
this leads to 
UnboundLocalError: local variable 'hepnames' referenced before assignment


there is another issue, that I just noticed. `get_hepnames()` returns a choice of names instead of `{}` when it doesn't find an exact match.  So instead of `if hepnames ...` the proper test is `if hepnames['have_hep'] ...`
https://github.com/inspirehep/invenio/blob/prod/modules/bibauthorid/lib/bibauthorid_webapi.py#L1470
clearly that isn't obvious from the function signature and you couldn't anticipate that

Reply:
Good points! I'll work on this.

Reply:
Done